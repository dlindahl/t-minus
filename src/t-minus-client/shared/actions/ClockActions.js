import ClockActions from '../actionTypes/ClockActionTypes';
import emptyObj from 'empty/object';
import formatDuration from '../util/formatDuration';
import partial from 'lodash/partial';
import present from 'present';

export const IntervalValue = 50;
let interval = null;

function calculateDuration(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      const timeRemaining = maxDuration - elapsedTime;
      const percentComplete = Math.min(elapsedTime / maxDuration, 1);
      return [formatDuration(timeRemaining), timeRemaining, percentComplete];
    default:
      return [formatDuration(elapsedTime), null, 1]
  }
}

function determineTimerElapsed(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      return elapsedTime > maxDuration;
    default:
      return false;
  }
}

export function clockTick(dispatch, getState) {
  const { clock } = getState();
  if(!clock.running) {
    return;
  }
  const currentTick = present();
  const delta = currentTick - clock.lastTick;
  const elapsedTime = clock.elapsedTime + delta;
  const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, elapsedTime, clock.timerValue);
  const hasTimerElapsed = determineTimerElapsed(clock.mode, elapsedTime, clock.timerValue);
  dispatch({
    type: ClockActions.CLOCK_TICK,
    payload: {
      currentTick,
      delta,
      duration,
      elapsedTime,
      hasTimerElapsed,
      mode: clock.mode,
      percentComplete,
      timeRemaining
    }
  });
}

export function pauseClock() {
  return {
    type: ClockActions.CLOCK_PAUSED,
    payload: emptyObj
  };
}

export function resetClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    clearInterval(interval);
    interval = null;
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_RESET,
      payload: {
        duration,
        timeRemaining,
        timerValue: clock.timerValue,
        percentComplete
      }
    });
  }
}

export function startClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, clock.elapsedTime, clock.timerDuration);
    dispatch({
      type: ClockActions.CLOCK_STARTED,
      payload: {
        duration,
        percentComplete,
        startedAt: present(),
        timeRemaining
      }
    });
    interval = setInterval(partial(clockTick, dispatch, getState), IntervalValue);
  };
}

export function changeTimerValue(timerValue) {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, timerValue);
    return dispatch({
      type: ClockActions.CLOCK_TIMER_CHANGED,
      payload: { duration, percentComplete, timeRemaining, timerValue }
    });
  }
}

export function stopwatchMode() {
  const mode = ClockActions.CLOCK_MODE_STOPWATCH;
  const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0);
  return {
    type: ClockActions.CLOCK_MODE_CHANGED,
    payload: {
      duration,
      mode,
      percentComplete,
      timeRemaining
    }
  }
}

export function timerMode() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const mode = ClockActions.CLOCK_MODE_TIMER;
    const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_MODE_CHANGED,
      payload: {
        duration,
        mode,
        percentComplete,
        timeRemaining
      }
    });
  }
}

export function toggleClock() {
  return (dispatch, getState) => {
    const clock = getState().clock
    if(clock.startedAt == null) {
      startClock()(dispatch, getState);
    } else if(clock.running) {
      dispatch(pauseClock());
    } else {
      dispatch(unpauseClock());
    }
  };
}

export function unpauseClock() {
  return {
    type: ClockActions.CLOCK_UNPAUSED,
    payload: {}
  };
}
