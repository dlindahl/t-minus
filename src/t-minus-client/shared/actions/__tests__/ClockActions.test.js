import * as clock from '../ClockActions'
import configureStore from 'redux-mock-store'
import emptyObj from 'empty/object'
import lolex from 'lolex'
import { merge } from 'lodash'

const STATES = {
  stopwatch: {
    clock: {
      elapsedTime: 0,
      lastTick: 0,
      mode: 'CLOCK_MODE_STOPWATCH',
      running: true,
      startedAt: 0,
      timerValue: 5000
    }
  },
  timer: {
    clock: {
      elapsedTime: 0,
      lastTick: 0,
      mode: 'CLOCK_MODE_TIMER',
      running: true,
      startedAt: 0,
      timerDuration: 5000,
      timerValue: 5000
    }
  }
}
const mockStore = configureStore()

describe('ClockActions', () => {
  describe('#pauseClock', () => {
    it('returns a payload', () => {
      const result = clock.pauseClock()
      expect(result.type).toBe('CLOCK_PAUSED')
      expect(result.payload).toBe(emptyObj)
    })
  })

  describe('#resetClock', () => {
    it('resets the clock', () => {
      const mockClock = lolex.install()
      const store = mockStore(STATES.stopwatch)
      clock.resetClock()(store.dispatch, store.getState)
      let actions = store.getActions()
      expect(actions.length).toBe(1)
      expect(actions[0].type).toBe('CLOCK_RESET')
      expect(actions[0].payload).toMatchSnapshot()
      mockClock.tick(clock.IntervalValue)
      actions = store.getActions()
      expect(actions.length).toBe(1)
      mockClock.uninstall()
    })

    it('resets the timer', () => {
      const mockClock = lolex.install()
      const store = mockStore(STATES.timer)
      clock.resetClock()(store.dispatch, store.getState)
      let actions = store.getActions()
      expect(actions[0].type).toBe('CLOCK_RESET')
      expect(actions[0].payload).toMatchSnapshot()
      mockClock.tick(clock.IntervalValue)
      actions = store.getActions()
      expect(actions.length).toBe(1)
      mockClock.uninstall()
    })
  })

  describe('#startClock', () => {
    it('starts the stopwatch', () => {
      const mockClock = lolex.install()
      const store = mockStore(STATES.stopwatch)
      clock.startClock()(store.dispatch, store.getState)
      let actions = store.getActions()
      expect(actions.length).toBe(1)
      expect(actions[0].type).toBe('CLOCK_STARTED')
      expect(actions[0].payload).toMatchSnapshot()
      mockClock.tick(clock.IntervalValue)
      actions = store.getActions()
      expect(actions.length).toBe(2)
      expect(actions[1].type).toBe('CLOCK_TICK')
      expect(actions[1].payload).toMatchSnapshot()
      mockClock.uninstall()
    })

    it('starts the timer', () => {
      const mockClock = lolex.install()
      const store = mockStore(STATES.timer)
      clock.startClock()(store.dispatch, store.getState)
      let actions = store.getActions()
      expect(actions[0].type).toBe('CLOCK_STARTED')
      expect(actions[0].payload).toMatchSnapshot()
      mockClock.tick(clock.IntervalValue)
      actions = store.getActions()
      expect(actions.length).toBe(2)
      expect(actions[1].type).toBe('CLOCK_TICK')
      expect(actions[1].payload).toMatchSnapshot()
      mockClock.uninstall()
    })
  })

  describe('#changeTimerValue', () => {
    it('resets and updates the timer value when in stopwatch mode', () => {
      const store = mockStore(STATES.stopwatch)
      clock.changeTimerValue()(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_TIMER_CHANGED')
      expect(action.payload).toMatchSnapshot()
    })

    it('resets and updates the timer value when in timer mode', () => {
      const store = mockStore(STATES.timer)
      clock.changeTimerValue(STATES.timer.clock.timerValue)(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_TIMER_CHANGED')
      expect(action.payload).toMatchSnapshot()
    })
  })

  describe('#stopwatchMode', () => {
    it('resets the clock and changes the mode from stopwatch to stopwatch', () => {
      const store = mockStore(STATES.stopwatch)
      store.dispatch(clock.stopwatchMode())
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_MODE_CHANGED')
      expect(action.payload).toMatchSnapshot()
    })

    it('resets the clock and changes the mode from timer to stopwatch', () => {
      const store = mockStore(STATES.timer)
      store.dispatch(clock.stopwatchMode())
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_MODE_CHANGED')
      expect(action.payload).toMatchSnapshot()
    })
  })

  describe('#timerMode', () => {
    it('resets the clock and changes the mode from stopwatch to timer', () => {
      const store = mockStore(STATES.stopwatch)
      clock.timerMode()(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_MODE_CHANGED')
      expect(action.payload).toMatchSnapshot()
    })

    it('resets the clock and changes the mode from timer to timer', () => {
      const store = mockStore(STATES.timer)
      clock.timerMode()(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_MODE_CHANGED')
      expect(action.payload).toMatchSnapshot()
    })
  })

  describe('#toggleClock', () => {
    it('starts the clock when not previously started', () => {
      const mockClock = lolex.install()
      const store = mockStore(merge({}, STATES.stopwatch, {
        clock: {
          running: false,
          startedAt: null
        }
      }))
      clock.toggleClock()(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_STARTED')
      expect(action.payload).toMatchSnapshot()
      mockClock.uninstall()
    })

    it('pauses the clock when running', () => {
      const mockClock = lolex.install()
      const store = mockStore(STATES.stopwatch)
      clock.toggleClock()(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_PAUSED')
      expect(action.payload).toMatchSnapshot()
      mockClock.uninstall()
    })

    it('unpauses the clock when paused', () => {
      const store = mockStore(merge({}, STATES.stopwatch, {
        clock: {
          running: false
        }
      }))
      clock.toggleClock()(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_UNPAUSED')
      expect(action.payload).toMatchSnapshot()
    })
  })

  describe('#clockTick', () => {
    it('does nothing if the clock is not running', () => {
      const store = mockStore(merge({}, STATES.stopwatch, {
        clock: {
          running: false
        }
      }))
      clock.clockTick(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action).not.toBeDefined()
    })

    it('recalculates the clock state when running', () => {
      const mockClock = lolex.install()
      mockClock.tick(clock.IntervalValue * 20)
      const store = mockStore(STATES.stopwatch)
      mockClock.tick(clock.IntervalValue * 20)
      clock.clockTick(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.payload).toMatchSnapshot()
      mockClock.uninstall()
    })

    it('recalculates the timer state when running', () => {
      const mockClock = lolex.install()
      mockClock.tick(clock.IntervalValue * 20)
      const store = mockStore(STATES.timer)
      mockClock.tick(clock.IntervalValue * 20)
      clock.clockTick(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.payload).toMatchSnapshot()
      mockClock.uninstall()
    })

    it('determines when the timer has expired', () => {
      const mockClock = lolex.install()
      mockClock.tick(clock.IntervalValue * 20)
      const store = mockStore(merge({}, STATES.timer, {
        clock: {
          timerDuration: 500,
          timerValue: 500
        }
      }))
      mockClock.tick(clock.IntervalValue * 20)
      clock.clockTick(store.dispatch, store.getState)
      const action = store.getActions()[0]
      expect(action.payload.hasTimerElapsed).toBeTruthy()
      mockClock.uninstall()
    })
  })

  describe('#pauseClock', () => {
    it('pauses the clock', () => {
      const store = mockStore(STATES.stopwatch)
      store.dispatch(clock.pauseClock())
      const action = store.getActions()[0]
      expect(action.type).toBe('CLOCK_PAUSED')
      expect(action.payload).toMatchSnapshot()
    })
  })
})
