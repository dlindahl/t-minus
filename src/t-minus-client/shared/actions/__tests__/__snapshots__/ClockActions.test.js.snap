exports[`ClockActions #changeTimerValue resets and updates the timer value when in stopwatch mode 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 0,
  },
  "percentComplete": 1,
  "timeRemaining": null,
  "timerValue": undefined,
}
`;

exports[`ClockActions #changeTimerValue resets and updates the timer value when in timer mode 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 5,
  },
  "percentComplete": 0,
  "timeRemaining": 5000,
  "timerValue": 5000,
}
`;

exports[`ClockActions #clockTick recalculates the clock state when running 1`] = `
Object {
  "currentTick": 2000,
  "delta": 2000,
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 2,
  },
  "elapsedTime": 2000,
  "hasTimerElapsed": false,
  "mode": "CLOCK_MODE_STOPWATCH",
  "percentComplete": 1,
  "timeRemaining": null,
}
`;

exports[`ClockActions #clockTick recalculates the timer state when running 1`] = `
Object {
  "currentTick": 2000,
  "delta": 2000,
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 3,
  },
  "elapsedTime": 2000,
  "hasTimerElapsed": false,
  "mode": "CLOCK_MODE_TIMER",
  "percentComplete": 0.4,
  "timeRemaining": 3000,
}
`;

exports[`ClockActions #pauseClock pauses the clock 1`] = `Object {}`;

exports[`ClockActions #resetClock resets the clock 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 0,
  },
  "percentComplete": 1,
  "timeRemaining": null,
  "timerValue": 5000,
}
`;

exports[`ClockActions #resetClock resets the timer 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 5,
  },
  "percentComplete": 0,
  "timeRemaining": 5000,
  "timerValue": 5000,
}
`;

exports[`ClockActions #startClock starts the stopwatch 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 0,
  },
  "percentComplete": 1,
  "startedAt": 0,
  "timeRemaining": null,
}
`;

exports[`ClockActions #startClock starts the stopwatch 2`] = `
Object {
  "currentTick": 50,
  "delta": 50,
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 50,
    "s": 0,
  },
  "elapsedTime": 50,
  "hasTimerElapsed": false,
  "mode": "CLOCK_MODE_STOPWATCH",
  "percentComplete": 1,
  "timeRemaining": null,
}
`;

exports[`ClockActions #startClock starts the timer 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 5,
  },
  "percentComplete": 0,
  "startedAt": 0,
  "timeRemaining": 5000,
}
`;

exports[`ClockActions #startClock starts the timer 2`] = `
Object {
  "currentTick": 50,
  "delta": 50,
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 950,
    "s": 4,
  },
  "elapsedTime": 50,
  "hasTimerElapsed": false,
  "mode": "CLOCK_MODE_TIMER",
  "percentComplete": 0.01,
  "timeRemaining": 4950,
}
`;

exports[`ClockActions #stopwatchMode resets the clock and changes the mode from stopwatch to stopwatch 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 0,
  },
  "mode": "CLOCK_MODE_STOPWATCH",
  "percentComplete": 1,
  "timeRemaining": null,
}
`;

exports[`ClockActions #stopwatchMode resets the clock and changes the mode from timer to stopwatch 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 0,
  },
  "mode": "CLOCK_MODE_STOPWATCH",
  "percentComplete": 1,
  "timeRemaining": null,
}
`;

exports[`ClockActions #timerMode resets the clock and changes the mode from stopwatch to timer 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 5,
  },
  "mode": "CLOCK_MODE_TIMER",
  "percentComplete": 0,
  "timeRemaining": 5000,
}
`;

exports[`ClockActions #timerMode resets the clock and changes the mode from timer to timer 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 5,
  },
  "mode": "CLOCK_MODE_TIMER",
  "percentComplete": 0,
  "timeRemaining": 5000,
}
`;

exports[`ClockActions #toggleClock pauses the clock when running 1`] = `Object {}`;

exports[`ClockActions #toggleClock starts the clock when not previously started 1`] = `
Object {
  "duration": Object {
    "d": 0,
    "h": 0,
    "m": 0,
    "ms": 0,
    "s": 0,
  },
  "percentComplete": 1,
  "startedAt": 0,
  "timeRemaining": null,
}
`;

exports[`ClockActions #toggleClock unpauses the clock when paused 1`] = `Object {}`;
